File: llms-txt/scripts/tree_gen.py
#!/usr/bin/env python3
"""
Cross-platform directory tree generator for llms.txt.

Generates an ASCII tree of the project structure, respecting .gitignore patterns
and skipping common generated/binary directories. Designed to be run by an AI agent
to avoid spending LLM tokens on mechanical file listing.

Usage:
    python3 tree_gen.py <project_root> [--max-depth N] [--max-files N]

Output:
    ASCII tree printed to stdout. Pipe or capture as needed.
"""

import argparse
import fnmatch
import os
import sys
from pathlib import Path

# Directories to always skip (case-sensitive match on directory name)
ALWAYS_SKIP_DIRS = {
    ".git",
    ".svn",
    ".hg",
    "node_modules",
    "__pycache__",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
    ".tox",
    ".nox",
    ".venv",
    "venv",
    "env",
    ".env",
    "dist",
    "build",
    "out",
    ".next",
    ".nuxt",
    ".output",
    ".turbo",
    ".vercel",
    ".netlify",
    "coverage",
    ".nyc_output",
    ".cache",
    ".parcel-cache",
    "target",          # Rust/Java build output
    "vendor",          # Go/PHP vendor
    ".gradle",
    ".idea",
    ".vscode",
    ".DS_Store",
    "Thumbs.db",
    ".terraform",
    ".serverless",
    "egg-info",
    "*.egg-info",
    ".eggs",
    ".ipynb_checkpoints",
}

# File patterns to always skip
ALWAYS_SKIP_FILES = {
    ".DS_Store",
    "Thumbs.db",
    "desktop.ini",
    "*.pyc",
    "*.pyo",
    "*.class",
    "*.o",
    "*.so",
    "*.dylib",
    "*.dll",
    "*.exe",
    "*.wasm",
    "*.min.js",
    "*.min.css",
    "*.map",
    "*.lock",
    "package-lock.json",
    "yarn.lock",
    "pnpm-lock.yaml",
    "uv.lock",
    "poetry.lock",
    "Gemfile.lock",
    "Cargo.lock",
    "composer.lock",
    "go.sum",
}


def parse_gitignore(project_root: Path) -> list[str]:
    """Parse .gitignore and return a list of patterns."""
    gitignore_path = project_root / ".gitignore"
    patterns = []
    if gitignore_path.is_file():
        try:
            with open(gitignore_path, "r", encoding="utf-8", errors="replace") as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith("#"):
                        patterns.append(line)
        except OSError:
            pass
    return patterns


def is_gitignored(rel_path: str, is_dir: bool, patterns: list[str]) -> bool:
    """Check if a relative path matches any gitignore pattern."""
    name = os.path.basename(rel_path)
    check_paths = [rel_path, name]
    if is_dir:
        check_paths.extend([rel_path + "/", name + "/"])

    for pattern in patterns:
        clean = pattern.rstrip("/")
        for cp in check_paths:
            if fnmatch.fnmatch(cp, clean) or fnmatch.fnmatch(cp, pattern):
                return True
            # Handle patterns like "dir/" matching directory names
            if is_dir and pattern.endswith("/") and fnmatch.fnmatch(name, clean):
                return True
            # Handle patterns with leading slash (relative to root)
            if pattern.startswith("/") and fnmatch.fnmatch(cp, pattern.lstrip("/")):
                return True
    return False


def should_skip_dir(name: str) -> bool:
    """Check if a directory name matches the always-skip list."""
    if name in ALWAYS_SKIP_DIRS:
        return True
    # Check wildcard patterns in skip list
    for pattern in ALWAYS_SKIP_DIRS:
        if "*" in pattern and fnmatch.fnmatch(name, pattern):
            return True
    return False


def should_skip_file(name: str) -> bool:
    """Check if a file name matches the always-skip list."""
    if name in ALWAYS_SKIP_FILES:
        return True
    for pattern in ALWAYS_SKIP_FILES:
        if "*" in pattern and fnmatch.fnmatch(name, pattern):
            return True
    return False


def generate_tree(
    root: Path,
    gitignore_patterns: list[str],
    max_depth: int = 10,
    max_files: int = 1000,
) -> str:
    """Generate an ASCII directory tree."""
    lines: list[str] = []
    file_count = 0

    lines.append(root.name + "/")

    def _walk(current: Path, prefix: str, depth: int):
        nonlocal file_count

        if depth > max_depth:
            lines.append(prefix + "└── ... (max depth reached)")
            return

        if file_count >= max_files:
            lines.append(prefix + "└── ... (max file count reached)")
            return

        try:
            entries = sorted(current.iterdir(), key=lambda e: (not e.is_dir(), e.name.lower()))
        except PermissionError:
            lines.append(prefix + "└── [permission denied]")
            return

        # Filter entries
        visible = []
        for entry in entries:
            name = entry.name
            rel = str(entry.relative_to(root))

            if entry.is_dir():
                if should_skip_dir(name):
                    continue
                if is_gitignored(rel, True, gitignore_patterns):
                    continue
                visible.append(entry)
            else:
                if should_skip_file(name):
                    continue
                if is_gitignored(rel, False, gitignore_patterns):
                    continue
                visible.append(entry)

        for i, entry in enumerate(visible):
            is_last = i == len(visible) - 1
            connector = "└── " if is_last else "├── "
            extension = "    " if is_last else "│   "

            if entry.is_dir():
                lines.append(f"{prefix}{connector}{entry.name}/")
                _walk(entry, prefix + extension, depth + 1)
            else:
                file_count += 1
                if file_count > max_files:
                    lines.append(f"{prefix}{connector}... (max file count reached)")
                    return
                lines.append(f"{prefix}{connector}{entry.name}")

    _walk(root, "", 0)
    return "\n".join(lines)


def main():
    parser = argparse.ArgumentParser(
        description="Generate an ASCII directory tree for llms.txt"
    )
    parser.add_argument("root", help="Path to the project root directory")
    parser.add_argument(
        "--max-depth",
        type=int,
        default=10,
        help="Maximum directory depth (default: 10)",
    )
    parser.add_argument(
        "--max-files",
        type=int,
        default=1000,
        help="Maximum number of files to list (default: 1000)",
    )
    args = parser.parse_args()

    root = Path(args.root).resolve()
    if not root.is_dir():
        print(f"Error: '{root}' is not a directory.", file=sys.stderr)
        sys.exit(1)

    gitignore_patterns = parse_gitignore(root)
    tree = generate_tree(root, gitignore_patterns, args.max_depth, args.max_files)
    print(tree)


if __name__ == "__main__":
    main()

File: llms-txt/scripts/tree_gen.js
#!/usr/bin/env node
/**
 * Cross-platform directory tree generator for llms.txt.
 *
 * Generates an ASCII tree of the project structure, respecting .gitignore patterns
 * and skipping common generated/binary directories. Designed to be run by an AI agent
 * to avoid spending LLM tokens on mechanical file listing.
 *
 * Usage:
 *     node tree_gen.js <project_root> [--max-depth N] [--max-files N]
 *
 * Output:
 *     ASCII tree printed to stdout. Pipe or capture as needed.
 */

const fs = require('fs');
const path = require('path');

// Directories to always skip (case-sensitive match on directory name)
const ALWAYS_SKIP_DIRS = new Set([
  '.git', '.svn', '.hg', 'node_modules', '__pycache__', '.mypy_cache',
  '.pytest_cache', '.ruff_cache', '.tox', '.nox', '.venv', 'venv', 'env',
  '.env', 'dist', 'build', 'out', '.next', '.nuxt', '.output', '.turbo',
  '.vercel', '.netlify', 'coverage', '.nyc_output', '.cache',
  '.parcel-cache', 'target', 'vendor', '.gradle', '.idea', '.vscode',
  '.DS_Store', 'Thumbs.db', '.terraform', '.serverless', 'egg-info',
  '*.egg-info', '.eggs', '.ipynb_checkpoints'
]);

// File patterns to always skip
const ALWAYS_SKIP_FILES = new Set([
  '.DS_Store', 'Thumbs.db', 'desktop.ini', '*.pyc', '*.pyo', '*.class',
  '*.o', '*.so', '*.dylib', '*.dll', '*.exe', '*.wasm', '*.min.js',
  '*.min.css', '*.map', '*.lock', 'package-lock.json', 'yarn.lock',
  'pnpm-lock.yaml', 'uv.lock', 'poetry.lock', 'Gemfile.lock', 'Cargo.lock',
  'composer.lock', 'go.sum'
]);

// Basic fnmatch implementation since Node lacks a built-in one
function fnmatch(name, pattern) {
  // Very basic globbing for standard gitignore/skip list patterns
  if (pattern.startsWith('*') && pattern.endsWith('*')) {
    return name.includes(pattern.slice(1, -1));
  }
  if (pattern.startsWith('*')) {
    return name.endsWith(pattern.slice(1));
  }
  if (pattern.endsWith('*')) {
    return name.startsWith(pattern.slice(0, -1));
  }
  return name === pattern;
}

function parseGitignore(projectRoot) {
  const gitignorePath = path.join(projectRoot, '.gitignore');
  const patterns = [];
  try {
    if (fs.existsSync(gitignorePath)) {
      const content = fs.readFileSync(gitignorePath, 'utf8');
      const lines = content.split('\n');
      for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed && !trimmed.startsWith('#')) {
          patterns.push(trimmed);
        }
      }
    }
  } catch (err) {
    // Ignore errors reading gitignore
  }
  return patterns;
}

function isGitignored(relPath, isDir, patterns) {
  const name = path.basename(relPath);
  // Normalise paths to forward slashes for matching
  const normalizedRelPath = relPath.replace(/\\/g, '/');
  
  const checkPaths = [normalizedRelPath, name];
  if (isDir) {
    checkPaths.push(normalizedRelPath + '/');
    checkPaths.push(name + '/');
  }

  for (const pattern of patterns) {
    const clean = pattern.replace(/\/$/, ''); // Remove trailing slash
    
    for (const cp of checkPaths) {
      if (fnmatch(cp, clean) || fnmatch(cp, pattern)) {
        return true;
      }
      
      if (isDir && pattern.endsWith('/') && fnmatch(name, clean)) {
         return true;
      }
      
      if (pattern.startsWith('/') && fnmatch(cp, pattern.replace(/^\//, ''))) {
          return true;
      }
    }
  }
  return false;
}

function shouldSkipDir(name) {
  if (ALWAYS_SKIP_DIRS.has(name)) return true;
  for (const pattern of ALWAYS_SKIP_DIRS) {
    if (pattern.includes('*') && fnmatch(name, pattern)) return true;
  }
  return false;
}

function shouldSkipFile(name) {
  if (ALWAYS_SKIP_FILES.has(name)) return true;
  for (const pattern of ALWAYS_SKIP_FILES) {
    if (pattern.includes('*') && fnmatch(name, pattern)) return true;
  }
  return false;
}

function generateTree(root, gitignorePatterns, maxDepth = 10, maxFiles = 1000) {
  const lines = [];
  let fileCount = 0;

  lines.push(path.basename(root) + '/');

  function walk(current, prefix, depth) {
    if (depth > maxDepth) {
      lines.push(prefix + '└── ... (max depth reached)');
      return;
    }

    if (fileCount >= maxFiles) {
      lines.push(prefix + '└── ... (max file count reached)');
      return;
    }

    let entries;
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch (err) {
      lines.push(prefix + '└── [permission denied]');
      return;
    }

    // Sort: directories first, then alphabetically
    entries.sort((a, b) => {
      const aIsDir = a.isDirectory();
      const bIsDir = b.isDirectory();
      if (aIsDir && !bIsDir) return -1;
      if (!aIsDir && bIsDir) return 1;
      return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
    });

    const visible = [];
    for (const entry of entries) {
      const name = entry.name;
      const fullPath = path.join(current, name);
      const relPath = path.relative(root, fullPath);
      const isDir = entry.isDirectory();

      if (isDir) {
        if (shouldSkipDir(name)) continue;
        if (isGitignored(relPath, true, gitignorePatterns)) continue;
        visible.push(entry);
      } else {
        if (shouldSkipFile(name)) continue;
        if (isGitignored(relPath, false, gitignorePatterns)) continue;
        visible.push(entry);
      }
    }

    for (let i = 0; i < visible.length; i++) {
      const entry = visible[i];
      const isLast = i === visible.length - 1;
      const connector = isLast ? '└── ' : '├── ';
      const extension = isLast ? '    ' : '│   ';
      const fullPath = path.join(current, entry.name);

      if (entry.isDirectory()) {
        lines.push(`${prefix}${connector}${entry.name}/`);
        walk(fullPath, prefix + extension, depth + 1);
      } else {
        fileCount++;
        if (fileCount > maxFiles) {
           lines.push(`${prefix}${connector}... (max file count reached)`);
           return;
        }
        lines.push(`${prefix}${connector}${entry.name}`);
      }
    }
  }

  walk(root, '', 0);
  return lines.join('\n');
}

function main() {
  const args = process.argv.slice(2);
  if (args.length === 0 || args[0] === '-h' || args[0] === '--help') {
    console.error('Usage: node tree_gen.js <project_root> [--max-depth N] [--max-files N]');
    process.exit(1);
  }

  let rootPath = args[0];
  let maxDepth = 10;
  let maxFiles = 1000;

  for (let i = 1; i < args.length; i++) {
    if (args[i] === '--max-depth' && i + 1 < args.length) {
      maxDepth = parseInt(args[++i], 10);
    } else if (args[i] === '--max-files' && i + 1 < args.length) {
      maxFiles = parseInt(args[++i], 10);
    }
  }

  const root = path.resolve(rootPath);
  try {
    if (!fs.statSync(root).isDirectory()) {
      console.error(`Error: '${root}' is not a directory.`);
      process.exit(1);
    }
  } catch(err) {
      console.error(`Error: Cannot access '${root}'.`);
      process.exit(1);
  }

  const gitignorePatterns = parseGitignore(root);
  const tree = generateTree(root, gitignorePatterns, maxDepth, maxFiles);
  console.log(tree);
}

if (require.main === module) {
  main();
}
